halo2-base/src/safe_types/mod.rs

```diff
@@ -0,0 +1,133 @@
+pub use crate::{
+    gates::{
+        flex_gate::GateInstructions,
+        range::{RangeChip, RangeInstructions},
+    },
+    utils::ScalarField,
+    AssignedValue, Context,
+    QuantumCell::{self, Constant, Existing, Witness},
+};
+use std::cmp::{max, min};
+
+type RawAssignedValues<F> = Vec<AssignedValue<F>>;
+
+const BITS_PER_BYTE: usize = 8;
+// Each AssignedValue can at most represent 8 bytes.
+const MAX_BYTE_PER_ELEMENT: usize = 8;
```

Comment:
this depends on `F`, probably better for future compability to use `F::CAPACITY / 8`

Reply:
To confirm, we don't have any preference on max bytes per element(e.g. 64bits), we just want to minimize the cells we need, right?

Reply:
Yes, and more like a safety check in case we add `F = Goldilocks` and someone tries to use 64 bits on Goldilocks field.

It looks like `MAX_BYTE_PER_ELEMENT` is only used to generate `byte_bases` and never used to check byte lengths anyways?